### PR TITLE
[MKLDNN] reorder the mem format for the AddBack mode in case src & dst is different

### PR DESCRIPTION
**Pull request by pengzhao-intel**
## Description ##
For MKL-DNN backend, the case is not covered that the src format is not as same as dst format for the AddBack mode. So the crash is observed in [example/reinforcement-learning/a3c](https://github.com/apache/incubator-mxnet/tree/master/example/reinforcement-learning/a3c). 

Therefore, the fix is to reorder the src format to dst format when the difference is detected.

The A3C case can pass with the fix. And verified the performance by [benchmark_score.py](https://github.com/apache/incubator-mxnet/blob/master/example/image-classification/benchmark_score.py), there's no performance effect. 

@zheng-da @TaoLv 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_base.cc [hunk 1]
```diff
@@ -128,8 +128,14 @@ void CommitOutput(const NDArray &arr, const mkldnn_output_t &res) {
   if (res.first == CopyBack) {
     const_cast<NDArray &>(arr).CopyFrom(*res.second);
   } else if (res.first == AddBack) {
-    auto mem = arr.GetMKLDNNData(res.second->get_primitive_desc());
-    CHECK(mem != nullptr);
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_base.cc]: I wonder whether it works if we just use GetMKLDNNDataReorder?

**Comment by zheng-da** [comment]
could you please add a test case?

File: src/operator/nn/mkldnn/mkldnn_base.cc (hunk 1 above)
Comment (pengzhao-intel) [reply src/operator/nn/mkldnn/mkldnn_base.cc]: I think it can work but it will be a little confusion to call reorder API for all cases.

**Comment by pengzhao-intel** [comment]
The crash happens in the backward of convolution when updating the weights. 

It's not easy to create the python level test case w/ the different memory format of the ndarray.
 
Maybe we can add a CPP case along with @juliusshufan 's CPP cases.

**Pull request closed without merging**